src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -82,6 +87,23 @@ static CurlHandler()
             // Feature detection
             s_supportsSSL = (CurlFeatures.CURL_VERSION_SSL & s_curlVersionInfoData.features) != 0;
             s_supportsAutomaticDecompression = (CurlFeatures.CURL_VERSION_LIBZ & s_curlVersionInfoData.features) != 0;
+
+            // We can send any pointer, even null seems to work. But sending a non-null value, just in case
+            if (CurlSslBackend.AllowsSslCallback(s_curlVersionInfoData.ssl_version))
+            {
+                s_clientCertificatePolicy = ClientCertificatePolicy.Callback;
+                VerboseTrace("openSSL compatible");
+            }
+            else if (CurlSslBackend.AllowsClientCertCommonName(s_curlVersionInfoData.ssl_version))
```

Comment:
> How is the user supposed to facilitate it?

 libcurl  is expected to be installed on the user's machine.   On linux, the three  most common* flavors of libcurl are: libcurl with openssl backend, libcurl with gnutls backend and libcurl with nss backend. In theory, users could install any of them. Or even build his own libcurl with other backends.

On OSX, libcurl with darwinssl is available OOB but users can still go ahead and install other  versions of libcurl, for example to work [around the certificate issue](http://curl.haxx.se/mail/archive-2013-10/0036.html).

At the moment, CurlHandler does not insist on the flavor of libcurl.   This line of code is trying to identify the user's choice of libcurl so that we can pass the certificate details in the most optimal way.

> If they've loaded a cert from a PFX and don't have it in the OSX store, what's the story? Worse, what if they loaded a cert from a PFX, and have a certificate with the same CN value but from a different issuer? What happens then?

This is a problem. I highlighted it elsewhere too. This is a problem with NSS backend too (on linux). 
On OSX (mavericks onwards), we cannot do https connections with out of box libcurl unless the cert is present in keygen-access and is trusted.  Please see the [http://curl.haxx.se/mail/archive-2013-10/0036.html](http://curl.haxx.se/mail/archive-2013-10/0036.html) for more details.  

> If we don't have a good story here we shouldn't try this integration path.

Kindly note that for these two backends (nss and darwinssl), this is the behavior of libcurl *even if I pass a file using CURLOPT_SSLCERT*. For those two backends, libcurl reads the file only to get the 'nickname' (their terminology) and use it to search against the respective cert store.
With the above code, we are only avoiding an export of the file (in managed layer) and an immediate reimport (by libcurl to read the nickname). 
The end result will be the same irrespective of what approach we use.

But I agree that, as far as libcurl is concerned, the documented/recommended way of passing a client certificate is via a file. 


*based on availability of packages in the package repos